Restore Multi-Modal LLM Content Support with Backward Compatibility

meopedevts opened the pull request:
This pull request proposes restoring support for multi-modal LLM content while ensuring backward compatibility with previous library versions.  

### **Motivation**  
While running tests with the Gemini model, I noticed that the `MultiContent` field in the `ChatCompletionMessage` struct was commented out. This PR aims to reintroduce this functionality with adjustments to maintain compatibility.  

### **Changes Introduced**  
- **New Field:** Instead of bringing back `MultiContent`, a new field named `ContentValue` has been introduced.  
  - I'm open to renaming it, but any alternative should still include "Content" for clarity.  
  - `ContentValue` is of type `ChatContent`, which includes:  
    - `Content string`  
    - `MultiContent []ChatMessagePart`  

```go
type ChatContent struct {
	content      string
	multiContent []ChatMessagePart
}

type ChatCompletionMessage struct {
	Role string `json:"role"`

	Content string `json:"content,omitzero"`

	ContentValue ChatContent `json:"content,omitzero"`

	Refusal      string      `json:"refusal,omitempty"`
	FunctionCall *FunctionCall `json:"function_call,omitempty"`
	ToolCalls []ToolCall `json:"tool_calls,omitempty"`
	ToolCallID string `json:"tool_call_id,omitempty"`
}
```

- **Helper Functions:**  
  - `NewContent(content string) ChatContent`  
  - `NewMultiContent(parts []ChatMessagePart) ChatContent`  
  - These functions ensure correct struct usage and prevent user errors.  

- **Custom JSON Serialization:**  
  - A custom `json.Marshal` method for `ChatContent` ensures correct serialization by prioritizing `Content` if both fields are populated.  

```go
func (c ChatContent) MarshalJSON() ([]byte, error) {
	if len(c.multiContent) > 0 {
		return json.Marshal(c.multiContent)
	}

	return json.Marshal(c.content)
}
```

- **Handling `content` Field Conflict:**  
  - The existing `Content` field is now marked as **deprecated** to maintain backward compatibility.  
  - A custom `json.Marshal` implementation for `ChatCompletionMessage` was added to avoid Go's restriction on duplicate JSON tags:  
    - `"govet: structtag: struct field ContentValue repeats json tag 'content' also at Content"`  
  - This method uses a temporary struct to enforce JSON field ordering, but we could simplify it using `map[string]interface{}` if necessary.  

```go
// MarshalJSON is customized to handle the ambiguity between the deprecated field `Content`
// and the new field `ContentValue`, both of which map to the JSON tag "content".
// This ensures backward compatibility while allowing for future deprecation of `Content`.
func (c ChatCompletionMessage) MarshalJSON() ([]byte, error) {
	// Create a new struct to garanted a order
	type Temp struct {
		Role         string        `json:"role"`
		Content      any           `json:"content,omitzero"`
		Refusal      string        `json:"refusal,omitempty"`
		FunctionCall *FunctionCall `json:"function_call,omitempty"`
		ToolCalls    []ToolCall    `json:"tool_calls,omitempty"`
		ToolCallID   string        `json:"tool_call_id,omitempty"`
	}

	temp := Temp{
		Role:         c.Role,
		Refusal:      c.Refusal,
		FunctionCall: c.FunctionCall,
		ToolCalls:    c.ToolCalls,
		ToolCallID:   c.ToolCallID,
	}

	if c.Content != "" {
		temp.Content = c.Content
	} else {
		temp.Content = c.ContentValue
	}

	return json.Marshal(temp)
}
```

### **Testing**  
A comprehensive test suite was added to `chat_test.go` to validate:  
- Correct serialization of `ChatCompletionMessage` when using:  
  - The deprecated `Content` field  
  - `ContentValue.Content`  
  - `ContentValue.MultiContent`  
- A scenario where both `Content` and `ContentValue` are set:  
  - The deprecated `Content` field currently takes precedence (this behavior can be changed if needed).  


This is an initial proposal, and I’m open to feedback to further refine this approach.  

Thanks for reviewing! 🚀

reVrost commented:
Thank you @meopedevts for the thorough pull request 🔥 . Appreciate it!
I would like to propose a slightly alternative approach to what you have if you don't mind.
We could leave `Content` field as is and uncomment `MultiContent` as a way to support your MultiContent use case with the Gemini model.

We would still have your changes to MarshalJSON similar to what you've done however setting the value of the final Content request to **use `MultiContent` if it is not empty or to use the default `Content` string field otherwise**.

I think this approach is simpler without having the need to deprecate old Content field and additional constructor. This also lets the user to choose whether they want to use the public field`Content` field or `MultiContent` in their request object easily.
**WDYT**?

meopedevts commented:
Hey @reVrost thanks for the feedback! I like the simpler approach with MultiContent while keeping MarshalJSON.

The new constructors for ContentValue do add unnecessary complexity.

I'll make the adjustments and push the updates soon.

meopedevts commented:
Added a commit refining the changes:  
  - Removed `ChatContent`, `NewContent`, `NewMultiContent`, and `ContentValue`  
  - Re-enabled `MultiContent` field  
  - Updated `MarshalJSON` to prioritize `MultiContent` if set, otherwise fallback to `Content`  
  - Cleaned up and renamed related tests  
  
Let me know if any further adjustments are needed! 🚀

reVrost commented on chat.go:
```diff
@@ -256,6 +256,37 @@ type ChatCompletionMessage struct {
+		ToolCallID:   c.ToolCallID,
+	}
+
+	if c.Content != "" {
```
```suggestion
	if len(c.MultiContent) == 0 {
```
Minor suggestion; this is so that if both are empty it will default to an empty string content rather than the multi content array.
